tests/load/gradual_orchestrator.go

```diff
@@ -0,0 +1,277 @@
+		currTime := time.Now()
+
+		tps := computeTPS(prevConfirmed, currConfirmed, currTime.Sub(prevTime))
+		o.setMaxObservedTPS(tps)
+
+		// if max TPS target has been reached and we don't terminate, then continue here
+		// so we do not keep increasing the target TPS
+		if achievedTargetTPS && !o.config.Terminate {
+			o.log.Info(
+				"current network state",
+				zap.Uint64("current TPS", tps),
+				zap.Uint64("max observed TPS", o.maxObservedTPS.Load()),
+			)
+			continue
+		}
+
+		if tps >= currTargetTPS.Load() {
+			if currTargetTPS.Load() >= o.config.MaxTPS {
+				achievedTargetTPS = true
+				o.log.Info(
+					"max TPS target reached",
+					zap.Uint64("max TPS target", currTargetTPS.Load()),
+					zap.Uint64("average TPS", tps),
+				)
+				if o.config.Terminate {
+					o.log.Info("terminating orchestrator")
+					break // Case 2
+				}
+				o.log.Info("orchestrator will now continue running at max TPS")
+				continue
+			}
+			o.log.Info(
+				"increasing TPS",
+				zap.Uint64("previous target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Uint64("new target TPS", currTargetTPS.Load()+o.config.Step),
+			)
+			currTargetTPS.Add(o.config.Step)
+			attempts = 1
+		} else {
+			if attempts >= o.config.MaxAttempts {
+				o.log.Info(
+					"max attempts reached",
+					zap.Uint64("attempted target TPS", currTargetTPS.Load()),
+					zap.Uint64("number of attempts", attempts),
+				)
+				break // Case 3
+			}
+			o.log.Info(
+				"failed to reach target TPS, retrying",
+				zap.Uint64("current target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Uint64("attempt number", attempts),
+			)
+			attempts++
+		}
+
+		prevConfirmed = currConfirmed
+		prevTime = currTime
+	}
+
+	return achievedTargetTPS
+}
+
+// GetObservedIssued returns the max TPS the orchestrator observed
+func (o *GradualOrchestrator[T, U]) GetMaxObservedTPS() uint64 {
+	return o.maxObservedTPS.Load()
+}
+
+// start a goroutine to each issuer to continuously send transactions.
+// if an issuer errors, all other issuers will stop as well.
+func (o *GradualOrchestrator[T, U]) issueTxs(ctx context.Context, currTargetTPS *atomic.Uint64) {
+	for _, agent := range o.agents {
+		o.issuerGroup.Go(func() error {
+			for {
+				if ctx.Err() != nil {
+					return nil //nolint:nilerr
+				}
+				currTime := time.Now()
+				txsPerIssuer := uint64(math.Ceil(float64(currTargetTPS.Load())/float64(len(o.agents))) * o.config.TxRateMultiplier)
+				// always listen until listener context is cancelled, do not call agent.Listener.IssuingDone().
+				for range txsPerIssuer {
+					tx, err := agent.Issuer.GenerateAndIssueTx(ctx)
+					if err != nil {
+						return err
+					}
+					agent.Listener.RegisterIssued(tx)
+				}
+				diff := time.Second - time.Since(currTime)
+				if diff > 0 {
+					timer := time.NewTimer(diff)
+					select {
+					case <-ctx.Done():
+						timer.Stop()
+						return nil
+					case <-timer.C:
+					}
+				}
```

Comment:
nit could flip the if condition to reduce indentation

```suggestion
				if diff <= 0 {
					continue
				}
				timer := time.NewTimer(diff)
				select {
				case <-ctx.Done():
					timer.Stop()
					return nil
				case <-timer.C:
				}
```